soc/nordic/common/dmm.c

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_out to user_buffer
+    // If no:
+    // - dynamically allocate buffer in correct memory region, using for example mem_attr_heap_alloc()
+    // - copy user buffer contents into allocated buffer
+
+    // Return error if dynamic allocation fails
+
+    // Check if device memory region is cacheable
+    // If yes, writeback all cache lines associated with output buffer (either user or allocated)
+    // If no, no action is needed
+}
+
+int dmm_dma_buffer_out_release(void const * buffer_out)
+{
+    // Get memory region that specified device can perform DMA transfers from, using devicetree
+
+    // Check if output buffer is contained within memory area managed by dynamic memory allocator
+    // If yes, free the buffer
+    // If no, no action is needed
+}
+
+int dmm_dma_buffer_in_prepare(const struct device *dev, void const * user_buffer, void ** buffer_in)
+{
+    // Get memory region that specified device can perform DMA transfers to, using devicetree
+
+    // Check if provided user buffer is already in correct memory region
+    // If yes, assign buffer_in to user_buffer
+    // If no, dynamically allocate buffer in correct memory region, using for example mem_attr_heap_alloc()
+
```

Comment:
If the buffer is cacheable, invalidate it to clear potential dirty bits in the cache.

Reply:
why is it needed? As far as I know DMA does not respect cache line status before writing to memory, and the cache lines will be invalidated during `dmm_dma_buffer_in_release` call (which is to be invoked once the DMA fills the buffer)

Reply:
Let's assume a cache line is tagged for the *buffer_in, the cache line contains the value 0xdeadbeef and the whole line is marked as dirty. The *buffer_in in RAM contains 0x00000000. After this function is called, the DMA buffer starts. After 2 bytes are received: 0x12340000 another software thread reuses the same line in cache to load data from another address. This triggers a hardware write-back of the dirty line. The write-back overwrites the memory at *buffer_in to 0xdeadbeef. The DMA transfer progresses and when it finishes the *buffer_in in RAM contains 0xdead5678 instead of 0x12345678. This is a data corruption scenario that must be prevented.

To prevent the data corruption scenario I described, the potential cache line tagged for the input DMA buffer cannot have any dirty bits set. The most efficient way to clean all dirty bits is invalidating a cache line.

Reply:
isn't this scenario avoided if we allocate buffers at cache line boundary and pad buffer size to cache line size?

Reply:
Let me expand my scenario:

The DMA buffer is provided by the driver user (application) with code like this:

```
uint8_t buffer[4] ALIGNED(CACHE_LINE) PADDED(CACHE_LINE) IN_DMABLE_MEMORY_FOR(spi0);

*(uint32_t *)buffer = 0xdeadbeef;

spi_receive(&spi0, buffer, ...);
```

So the buffer is at the cache line boundary. Yet when the driver starts execution, the cache line can contain dirty data `0xdeadbeef`. To make sure this dirty line does not cause data corruption, it must be invalidated before DMA starts writing to the buffer.

Reply:
> To make sure this dirty line does not cause data corruption

How this dirty line can cause data corruption? When cache line is written-back after DMA finalizes the RX transfer but before cache line invalidation?

Reply:
> > To make sure this dirty line does not cause data corruption
> 
> How this dirty line can cause data corruption? When cache line is written-back after DMA finalizes the RX transfer but before cache line invalidation on RX event handling?

Yes. To be precise the data corruption happens if the HW initializes write-back automatically (cache line eviction) during a time window. The time window starts when the first data unit gets stored by DMA in the buffer, and it ends when the driver invalidates the line.

Reply:
added